node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,734 @@
+func (w *Watcher) handleObservationRequests(ctx context.Context, errC chan error, ce *connectionEntry) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case r := <-ce.obsvReqC:
+			if vaa.ChainID(r.ChainId) != ce.chainID {
+				panic("invalid chain ID")
+			}
+
+			reqTxHashStr := hex.EncodeToString(r.TxHash)
+			w.logger.Info("received observation request", zap.String("chain", ce.chainName), zap.String("txHash", reqTxHashStr))
+
+			client := &http.Client{
+				Timeout: time.Second * 5,
+			}
+
+			// Query for tx by hash.
+			resp, err := client.Get(fmt.Sprintf("%s/cosmos/tx/v1beta1/txs/%s", w.lcdUrl, reqTxHashStr))
+			if err != nil {
+				w.logger.Error("query tx response error", zap.String("chain", ce.chainName), zap.Error(err))
+				continue
+			}
+			txBody, err := io.ReadAll(resp.Body)
+			if err != nil {
+				w.logger.Error("query tx response read error", zap.String("chain", ce.chainName), zap.Error(err))
+				resp.Body.Close()
+				continue
+			}
+			resp.Body.Close()
+
+			txJSON := string(txBody)
+			txHashRaw := gjson.Get(txJSON, "tx_response.txhash")
+			if !txHashRaw.Exists() {
+				w.logger.Error("tx does not have tx hash", zap.String("chain", ce.chainName), zap.String("payload", txJSON))
+				continue
+			}
+			txHashStr := txHashRaw.String()
+			txHash, err := vaa.StringToHash(txHashStr)
+			if err != nil {
+				w.logger.Error("tx does not have tx hash", zap.String("chain", ce.chainName), zap.String("payload", txJSON))
+				continue
+			}
+
+			events := gjson.Get(txJSON, "tx_response.events")
+			if !events.Exists() {
+				w.logger.Error("tx has no events", zap.String("chain", ce.chainName), zap.String("payload", txJSON))
+				continue
+			}
+
+			for _, event := range events.Array() {
+				if !event.IsObject() {
+					w.logger.Error("event is invalid", zap.String("chain", ce.chainName), zap.String("tx_hash", txHashStr), zap.String("event", event.String()))
+					continue
+				}
+				eventType := gjson.Get(event.String(), "type")
+				if eventType.String() == "wasm" {
+					w.logger.Debug("found wasm event in reobservation", zap.String("chain", ce.chainName), zap.Stringer("txHash", txHash))
+					evt, err := parseIbcReceivePublishEvent(w.logger, w.contractAddress, event, txHash)
+					if err != nil {
+						w.logger.Error("failed to parse wasm event", zap.String("chain", ce.chainName), zap.Error(err), zap.Any("event", event))
+						continue
+					}
+
+					if evt != nil {
+						w.processIbcReceivePublishEvent(txHash, evt, "reobservation")
+					}
+				} else {
+					w.logger.Debug("ignoring uninteresting event in reobservation", zap.String("chain", ce.chainName), zap.Stringer("txHash", txHash), zap.String("eventType", eventType.String()))
+				}
+			}
+		}
+	}
+}
+
+// parseIbcReceivePublishEvent parses a wasm event. If it is from the desired contract and for the desired action, it returns an event. Otherwise, it returns nil.
```

Comment:
make more clear it returns an error if it's not from the desired contract but nil if it's a different event

Reply:
Done.